File: companion/bolumler/b10/ornek-01/cozum.py
import argparse
import math
from pathlib import Path
import platform

import numpy as np
import pandas as pd
import scipy
from scipy import stats


ARAMA_UST = 10000


def hazirla(veri, plan=False):
    sutunlar = (["hacim", "fark", "standart_sapma", "alfa", "hedef_guc"] if plan else
                ["hacim", "ortalama", "standart_sapma", "referans", "alfa"])
    if list(veri.columns) != sutunlar or len(veri) != 1:
        raise ValueError("Tek ozet satiri ve dogru sutun sirasi gerekli.")
    if veri.isna().any().any():
        raise ValueError("Eksik ozet kabul edilmez.")
    for sutun in sutunlar:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal ozet gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Sonlu ozet gerekli.")
    temiz = veri.reset_index(drop=True).copy(deep=True)
    hacim = float(temiz.loc[0, "hacim"])
    if hacim < 2 or not hacim.is_integer():
        raise ValueError("Hacim en az 2 olan tam sayi olmali.")
    if temiz.loc[0, "standart_sapma"] <= 0 or not 0 < temiz.loc[0, "alfa"] < 1:
        raise ValueError("Standart sapma pozitif, alfa 0 ile 1 arasinda olmali.")
    if plan and (temiz.loc[0, "fark"] == 0 or not temiz.loc[0, "alfa"] < temiz.loc[0, "hedef_guc"] < 1):
        raise ValueError("Hacim plani icin sifirdan farkli etki ve alfa < hedef guc < 1 gerekli.")
    return temiz


def guc_hesapla(hacim, etki, alfa):
    if not all(math.isfinite(float(value)) for value in [hacim, etki, alfa]):
        raise ValueError("Sonlu guc girdileri gerekli.")
    if hacim < 2 or not 0 < alfa < 1:
        raise ValueError("Guc icin hacim >= 2 ve 0 < alfa < 1 gerekli.")
    serbestlik = hacim - 1
    kritik = stats.t.ppf(1 - alfa / 2, df=serbestlik)
    merkezdisilik = abs(etki) * math.sqrt(hacim)
    guc = float(stats.nct.cdf(-kritik, df=serbestlik, nc=merkezdisilik) +
                stats.nct.sf(kritik, df=serbestlik, nc=merkezdisilik))
    if not math.isfinite(guc) or not 0 <= guc <= 1:
        raise ValueError("Guc sayisal olarak hesaplanamadi.")
    return guc


def en_kucuk_hacim(etki, alfa, hedef):
    if etki == 0 or not alfa < hedef < 1:
        raise ValueError("Sonlu hacim plani icin sifir olmayan etki ve alfa < hedef < 1 gerekli.")
    alt, ust = 2, 2
    while guc_hesapla(ust, etki, alfa) < hedef:
        if ust == ARAMA_UST:
            raise ValueError("Hedef guc 10000 gozleme kadar bulunamadi; bu paket arama siniri asildi.")
        ust = min(2 * ust, ARAMA_UST)
    while alt < ust:
        orta = (alt + ust) // 2
        if guc_hesapla(orta, etki, alfa) >= hedef:
            ust = orta
        else:
            alt = orta + 1
    return alt


def hesapla(veri, plan):
    veri, plan = hazirla(veri), hazirla(plan, plan=True)
    hacim, ortalama, sapma, referans, alfa = map(float, veri.iloc[0])
    plan_hacim, fark, plan_sapma, plan_alfa, hedef = map(float, plan.iloc[0])
    serbestlik = hacim - 1
    standart_hata = sapma / math.sqrt(hacim)
    t_degeri = (ortalama - referans) / standart_hata
    p_degeri = float(2 * stats.t.sf(abs(t_degeri), df=serbestlik))
    kritik = float(stats.t.ppf(1 - alfa / 2, df=serbestlik))
    hata_payi = kritik * standart_hata
    test = dict(serbestlik=serbestlik, standart_hata=standart_hata, t=t_degeri,
                p_cift=p_degeri, cohen_d=(ortalama - referans) / sapma, kritik_t=kritik,
                alt_sinir=ortalama - hata_payi, ust_sinir=ortalama + hata_payi,
                hata_payi=hata_payi, genislik=2 * hata_payi, reddet_cift=int(p_degeri < alfa))
    etki = fark / plan_sapma
    guc = guc_hesapla(plan_hacim, etki, plan_alfa)
    gereken = en_kucuk_hacim(etki, plan_alfa, hedef)
    guc_onceki = guc_hesapla(gereken - 1, etki, plan_alfa) if gereken > 2 else float("nan")
    tasarim = dict(plan_d=etki, plan_serbestlik=plan_hacim - 1,
                   plan_kritik_t=float(stats.t.ppf(1 - plan_alfa / 2, df=plan_hacim - 1)),
                   merkezdisilik=abs(etki) * math.sqrt(plan_hacim), guc=guc, beta=1 - guc,
                   gereken_hacim=gereken, onceki_guc=guc_onceki,
                   gereken_guc=guc_hesapla(gereken, etki, plan_alfa))
    rows = [("girdi", key, float(veri.loc[0, key])) for key in veri.columns]
    rows += [("plan_girdi", key, float(plan.loc[0, key])) for key in plan.columns]
    rows += [("test", key, value) for key, value in test.items()]
    rows += [("plan", key, value) for key, value in tasarim.items()]
    sonuc = pd.DataFrame(rows, columns=["degisken", "olcu", "deger"])
    degerler = sonuc.loc[sonuc.olcu != "onceki_guc", "deger"]
    if not np.isfinite(degerler).all():
        raise ValueError("Sonuc sayisal olarak hesaplanamadi.")
    return sonuc


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen.deger, errors="raise")
    if not np.isfinite(hedefler).all() or not np.isfinite(sonuc.deger).all():
        raise ValueError("Sonlu kontrol degerleri gerekli.")
    if not all(math.isclose(actual, expected, rel_tol=1e-9, abs_tol=1e-9)
               for actual, expected in zip(sonuc.deger, hedefler)):
        raise ValueError("Kontrol degerleri uyusmuyor.")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(plan):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    temiz = hazirla(plan, plan=True)
    hacim, fark, sapma, alfa, hedef = map(float, temiz.iloc[0])
    etki = fark / sapma
    gereken = en_kucuk_hacim(etki, alfa, hedef)
    ust = min(ARAMA_UST, max(60, hacim + 10, gereken + 10))
    hacimler = np.unique(np.rint(np.linspace(2, ust, 120)).astype(int))
    fig, panel = plt.subplots(figsize=(8, 4.5))
    for carpan, stil in [(1, "-"), (.5, "--")]:
        gucler = [guc_hesapla(int(aday), etki * carpan, alfa) for aday in hacimler]
        panel.plot(hacimler, gucler, stil, label=f"Planlanan |d| = {abs(etki * carpan):g}")
    panel.axhline(hedef, color="#555555", linestyle=":", label=f"Hedef guc = {hedef:g}")
    panel.scatter([gereken], [guc_hesapla(gereken, etki, alfa)], color="#087F5B", zorder=3)
    panel.annotate(f"En kucuk n = {gereken}", (gereken, guc_hesapla(gereken, etki, alfa)),
                   xytext=(8, -22), textcoords="offset points")
    panel.set(xlabel="Bagimsiz gozlem sayisi n", ylabel="Iki tarafli test gucu", ylim=(0, 1),
              title=f"Ileriye donuk tek orneklem t testi plani (alfa={alfa:g})")
    panel.legend(loc="lower right")
    fig.tight_layout()
    hedef_dizin = Path("ciktilar/python")
    hedef_dizin.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef_dizin / "guc-egrisi.png", dpi=150)
    plt.close(fig)
    print("Grafik: ciktilar/python/guc-egrisi.png")


def main():
    parser = argparse.ArgumentParser(description="B10: tek orneklem t testi ve ileriye donuk guc")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri, plan = pd.read_csv("veri.csv"), pd.read_csv("plan.csv")
    sonuc = hesapla(veri, plan)
    print(sonuc.to_string(index=False))
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(plan)
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}, SciPy {scipy.__version__}")


if __name__ == "__main__":
    main()

File: companion/bolumler/b10/ornek-01/cozum.R
ARAMA_UST <- 10000

hazirla <- function(veri, plan = FALSE) {
  sutunlar <- if (plan) c("hacim", "fark", "standart_sapma", "alfa", "hedef_guc") else
    c("hacim", "ortalama", "standart_sapma", "referans", "alfa")
  if (!identical(names(veri), sutunlar) || nrow(veri) != 1) stop("Tek ozet satiri ve dogru sutun sirasi gerekli.")
  if (!all(vapply(veri, is.numeric, logical(1))) || anyNA(veri) ||
      any(!is.finite(as.matrix(veri)))) stop("Sonlu sayisal ozet gerekli.")
  if (veri$hacim < 2 || veri$hacim != floor(veri$hacim)) stop("Hacim en az 2 olan tam sayi olmali.")
  if (veri$standart_sapma <= 0 || veri$alfa <= 0 || veri$alfa >= 1) stop("Sapma veya alfa gecersiz.")
  if (plan && (veri$fark == 0 || veri$hedef_guc <= veri$alfa || veri$hedef_guc >= 1)) {
    stop("Hacim plani icin sifir olmayan etki ve alfa < hedef guc < 1 gerekli.")
  }
  veri
}

guc_hesapla <- function(hacim, etki, alfa) {
  if (any(!is.finite(c(hacim, etki, alfa))) || hacim < 2 || alfa <= 0 || alfa >= 1) {
    stop("Guc girdileri gecersiz.")
  }
  guc <- power.t.test(n = hacim, delta = abs(etki), sd = 1, sig.level = alfa,
                      type = "one.sample", alternative = "two.sided", strict = TRUE)$power
  if (!is.finite(guc) || guc < 0 || guc > 1) stop("Guc sayisal olarak hesaplanamadi.")
  guc
}

en_kucuk_hacim <- function(etki, alfa, hedef) {
  if (etki == 0 || hedef <= alfa || hedef >= 1) stop("Hacim plani girdileri gecersiz.")
  alt <- 2
  ust <- 2
  while (guc_hesapla(ust, etki, alfa) < hedef) {
    if (ust == ARAMA_UST) stop("Hedef 10000 gozleme kadar bulunamadi; arama siniri asildi.")
    ust <- min(2 * ust, ARAMA_UST)
  }
  while (alt < ust) {
    orta <- floor((alt + ust) / 2)
    if (guc_hesapla(orta, etki, alfa) >= hedef) ust <- orta else alt <- orta + 1
  }
  alt
}

hesapla <- function(veri, plan) {
  veri <- hazirla(veri)
  plan <- hazirla(plan, plan = TRUE)
  serbestlik <- veri$hacim - 1
  standart_hata <- veri$standart_sapma / sqrt(veri$hacim)
  t_degeri <- (veri$ortalama - veri$referans) / standart_hata
  p_degeri <- 2 * pt(abs(t_degeri), df = serbestlik, lower.tail = FALSE)
  kritik <- qt(1 - veri$alfa / 2, df = serbestlik)
  hata_payi <- kritik * standart_hata
  test <- c(serbestlik = serbestlik, standart_hata = standart_hata, t = t_degeri,
             p_cift = p_degeri, cohen_d = (veri$ortalama - veri$referans) / veri$standart_sapma,
             kritik_t = kritik, alt_sinir = veri$ortalama - hata_payi,
             ust_sinir = veri$ortalama + hata_payi, hata_payi = hata_payi,
             genislik = 2 * hata_payi, reddet_cift = as.integer(p_degeri < veri$alfa))
  etki <- plan$fark / plan$standart_sapma
  guc <- guc_hesapla(plan$hacim, etki, plan$alfa)
  gereken <- en_kucuk_hacim(etki, plan$alfa, plan$hedef_guc)
  guc_onceki <- if (gereken > 2) guc_hesapla(gereken - 1, etki, plan$alfa) else NA_real_
  tasarim <- c(plan_d = etki, plan_serbestlik = plan$hacim - 1,
                plan_kritik_t = qt(1 - plan$alfa / 2, df = plan$hacim - 1),
                merkezdisilik = abs(etki) * sqrt(plan$hacim), guc = guc, beta = 1 - guc,
                gereken_hacim = gereken, onceki_guc = guc_onceki,
                gereken_guc = guc_hesapla(gereken, etki, plan$alfa))
  sonuc <- data.frame(degisken = c(rep("girdi", 5), rep("plan_girdi", 5), rep("test", 11), rep("plan", 9)),
                       olcu = c(names(veri), names(plan), names(test), names(tasarim)),
                       deger = c(unlist(veri), unlist(plan), test, tasarim), row.names = NULL)
  if (any(!is.finite(sonuc$deger[sonuc$olcu != "onceki_guc"]))) stop("Sonuc hesaplanamadi.")
  sonuc
}

kontrol_et <- function(sonuc, beklenen) {
  if (!identical(names(beklenen), c("degisken", "olcu", "deger")) ||
      !identical(sonuc$degisken, beklenen$degisken) || !identical(sonuc$olcu, beklenen$olcu)) {
    stop("Kontrol etiketleri veya sirasi uyusmuyor.")
  }
  if (!is.numeric(beklenen$deger) || any(!is.finite(beklenen$deger)) ||
      any(!is.finite(sonuc$deger))) stop("Sonlu kontrol degerleri gerekli.")
  tolerans <- pmax(1e-9, 1e-9 * pmax(abs(sonuc$deger), abs(beklenen$deger)))
  if (any(abs(sonuc$deger - beklenen$deger) > tolerans)) stop("Kontrol degerleri uyusmuyor.")
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}

grafik_kaydet <- function(plan) {
  plan <- hazirla(plan, plan = TRUE)
  etki <- plan$fark / plan$standart_sapma
  gereken <- en_kucuk_hacim(etki, plan$alfa, plan$hedef_guc)
  ust <- min(ARAMA_UST, max(60, plan$hacim + 10, gereken + 10))
  hacimler <- unique(round(seq(2, ust, length.out = 120)))
  gucler <- vapply(hacimler, guc_hesapla, numeric(1), etki = etki, alfa = plan$alfa)
  kucuk_gucler <- vapply(hacimler, guc_hesapla, numeric(1), etki = etki / 2, alfa = plan$alfa)
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/guc-egrisi.png", width = 1200, height = 675, res = 150)
  on.exit(dev.off())
  plot(hacimler, gucler, type = "l", col = "#1F4E79", ylim = c(0, 1),
       xlab = "Bagimsiz gozlem sayisi n", ylab = "Iki tarafli test gucu",
       main = paste("Ileriye donuk tek orneklem t testi plani; alfa =", plan$alfa))
  lines(hacimler, kucuk_gucler, lty = 2, col = "#087F5B")
  abline(h = plan$hedef_guc, lty = 3, col = "#555555")
  points(gereken, guc_hesapla(gereken, etki, plan$alfa), pch = 19)
  text(gereken, guc_hesapla(gereken, etki, plan$alfa) - .06, paste("En kucuk n =", gereken))
  legend("bottomright", legend = c(paste("Planlanan |d| =", abs(etki)),
         paste("Planlanan |d| =", abs(etki) / 2), paste("Hedef guc =", plan$hedef_guc)),
         col = c("#1F4E79", "#087F5B", "#555555"), lty = c(1, 2, 3), cex = .8)
  cat("Grafik: ciktilar/r/guc-egrisi.png\n")
}

secenek <- commandArgs(trailingOnly = TRUE)
if (any(!secenek %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek.")
veri <- read.csv("veri.csv", check.names = FALSE)
plan <- read.csv("plan.csv", check.names = FALSE)
sonuc <- hesapla(veri, plan)
print(sonuc, row.names = FALSE)
if ("--check" %in% secenek) kontrol_et(sonuc, read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE))
if ("--grafik" %in% secenek) grafik_kaydet(plan)
sessionInfo()